### DISCUSSION TITLE
Spacecraft mode management updates

### DISCUSSION THREAD

**Discussion by timcanham** [nasa/fprime#2622] [Ideas]
This is a follow-up to [this](https://github.com/nasa/fprime/discussions/2310) discussion.

Here is a proposal for how to do basic mode management:

1\) FPP updates - Add a new `SpacecraftMode` [enumeration](https://fprime-community.github.io/fpp/fpp-spec.html#Definitions_Enum-Definitions) that lives in the project [config](https://github.com/nasa/fprime/tree/devel/config) directory. That allows projects to customize it to modes they like. For instance:

```
enum SpacecraftMode {
  TEST,
  LAUNCH,
  ORBIT_IDLE,
  ORBIT_SCIENCE,
  SAFE
}
```

Add an optional ```modes``` attribute to [commands](https://fprime-community.github.io/fpp/fpp-spec.html#Specifiers_Command-Specifiers) that specify a set of modes where the command is valid. For instance:

```
@ A sync command with parameters
sync command SyncParams(
    param1: U32 @< Param 1
    param2: string @< Param 2
    modes: TEST, ORBIT_SCIENCE
) opcode 0x02
```

Define a port type that allows getting the current mode  (`SpacecraftModeGet`) for command checking, and setting (`SpacecraftModeSet`) for components (like fault protection) to set the mode.

2\) FPP Autocoder

For commands that have the `modes` attribute, add a call to generated code to check the spacecraft mode via the `SpacecraftModeGet` port prior to executing the command.

3\) Components: Write a basic `SpacecraftMode` component. This component would implement the `SpacecraftModeGet` and `SpacecraftModeSet` ports to store the current spacecraft mode. It would have commands to set the mode for ground operators.

If projects wanted more sophisticated mode management, they could implement their own component with `SpacecraftModeGet` and `SpacecraftModeSet` ports, but the above would provide a basic implementation.

This could also be wired into the basic [fault protection](https://github.com/nasa/fprime/discussions/2536) component as a mode switch due to a fault.

**Comment by abdelrahman47** [comment]
Thanks for the good proposal. 

I have implemented the Mode management component and it works well. 

I have a question regarding the port invocations. 

Is there a way to make an announcement to all ports?  or it should be invoked one by one since in F Prime an output port can only be connected to one input port.

**Reply by LeStarch** [reply]
Typically one would have an array of output ports, and loop over them to send each the same message.
